R/dependecies.R

```diff
@@ -0,0 +1,437 @@
+    ## for any version of the package.
+    if(!package$name %in% names(previously_parsed_deps)) {
+      previously_parsed_deps[[package$name]] <- list()
+    }
+    if (!paste(package$version) %in% 
+      names(previously_parsed_deps[[package$name]])) {
+      dependency_output <- get_dependencies(
+       structure(package
+         , class = c(package$remote %||% "CRAN"
+           , class(package)))
+       , lock)
+      package <- dependency_output$package
+      single_package_dependencies <- dependency_output$dependencies
+
+      ## Our package comes out of dependency search with a download_path
+      ## attached.  We will use this for installation (assuming we keep this
+      ## version of the package.
+      current_dependencies[[package$name]] <- package
+
+      ## Store the dependencies and corresponding package object in our humongous
+      ## previously_parsed_deps object using version as the second key level
+      previously_parsed_deps[[package$name]][[paste(package$version)]] <-
+        list(package = package, dependencies = single_package_dependencies)
+    } else {
+      single_package_dependencies <-
+        previously_parsed_deps[[package$name]][[
+          paste(package$version)]]$dependencies
+      current_dependencies[[package$name]] <-
+        previously_parsed_deps[[package$name]][[
+          paste(package$version)]]$package
+    }
+    ## Now combine the dependencies from this package with our big dependency
+    ## list.
+
+    current_dependencies <- combine_dependencies(
+      single_package_dependencies
+      , current_dependencies
+      , package$name)
+  }
+
+  if (identical(master_list, current_dependencies)) return(master_list)
+
+  ## Run through the entire list again, because we have a bunch of new packages
+  get_dependencies_for_list(current_dependencies, lock, previously_parsed_deps)
+}
+
+#' Attach the latest available lockbox version to a package
+add_latest_version_in_lockbox <- function(package) {
+  package$latest_version_in_lockbox <- max_package_version(
+    list.files(file.path(lockbox_library(), package)))
+  package
+}
+
+#' Take a vector of versions and find the max without coercing to package_version
+max_package_version <- function(versions) {
+  if (length(versions) == 0) return(NULL)
+  formatted_versions <- package_version(as.character(versions))
+  versions[which(formatted_versions == max(formatted_versions))[1]]
+}
+
+#' Check a dependency list for inclusion in the lockfile and replace the package
+#' with the locked version if it does appear there.  Also throw
+#' an error if we require a dependency version greater than that specified by
+#' the lockfile.
+replace_with_lock <- function(package, lock) {
+  lock_names <- vapply(lock, function(l) l$name, character(1))
+  if (package$name %in% lock_names) {
+    locked_package <- Filter(function(l) l$name == package$name, lock)[[1]]
+    if (!is.na(package$version) && package_version(as.character(package$version)) >
```

Comment:
Not for dependency packages, due to to package_version's incompatibility with NA versions.  Even if, instead of NA, we used some other version to represent "latest_version_available" we would still have problems with either conversion to package_version (if it doesn't recognize the format) or collision with actual versions (if it does)